File: lab_6/A3/src/sr/ice/client/IceClient.java
package sr.ice.client;

import Office.*;
import com.zeroc.Ice.*;

import java.io.IOException;
import java.lang.Exception;
import java.util.concurrent.CompletableFuture;

public class IceClient {
    public static void main(String[] args) {
        int status = 0;
        Communicator communicator = null;

        try {
            communicator = Util.initialize(args);

            ObjectPrx server = communicator.propertyToProxy("Clerk.Proxy");

            ClerkPrx clerk = ClerkPrx.checkedCast(server);
            if (clerk == null) throw new Error("Invalid proxy");

            ObjectAdapter adapter = communicator.createObjectAdapter("");
            ClientPrx clientPrx = ClientPrx.uncheckedCast(adapter.addWithUUID(new ClientI()));
            adapter.activate();
            clerk.ice_getConnection().setAdapter(adapter);

            CompletableFuture<Long> cfl = null;
            String line = null;
            java.io.BufferedReader in = new java.io.BufferedReader(new java.io.InputStreamReader(System.in));

            PersonData personData = new PersonData("Arkadiusz", "Cwikla", "505abc");

            clerk.sayHello(personData, clientPrx);

            do {
                try {
                    System.out.println("==> ");
                    line = in.readLine();
                    if (line.equals("reg")) {
                        long time = clerk.register(
                                personData,
                                new RegistrationRequest("Opel", (short) (2003), engineType.GAS)
                        );
                        System.out.printf("Time to wait: %s [ms]%n", time);
                    } else if (line.equals("ren")) {
                        licenseCategory[] categories = {licenseCategory.A, licenseCategory.B};

                        long time = clerk.renewLicense(
                                personData,
                                new RenewLicenseRequest(true, (short) 5, categories)
                        );
                        System.out.printf("Time to wait: %s [ms]%n", time);
                    } else if (line.equals("ren-bad")) {
                        licenseCategory[] categories = {licenseCategory.A, licenseCategory.B};

                        long time = clerk.renewLicense(
                                personData,
                                new RenewLicenseRequest(false, (short) 5, categories)
                        );
                        System.out.printf("Time to wait: %s [ms]%n", time);
                    } else if (line.equals("ins")) {
                        long time = clerk.instructorLicense(
                                personData,
                                new InstructorLicenseRequest(true, true, true)
                        );
                        System.out.printf("Time to wait: %s [ms]%n", time);
                    } else if (line.equals("ins-bad")) {
                        long time = clerk.instructorLicense(
                                personData,
                                new InstructorLicenseRequest(true, false, true)
                        );
                        System.out.printf("Time to wait: %s [ms]%n", time);
                    }
                } catch (IOException | TwowayOnlyException ex) {
                    System.err.println(ex);
                }
            }
            while (!line.equals("x"));


        } catch (LocalException e) {
            e.printStackTrace();
            status = 1;
        } catch (Exception e) {
            System.err.println(e.getMessage());
            status = 1;
        }
        if (communicator != null) { //clean
            try {
                communicator.destroy();
            } catch (Exception e) {
                System.err.println(e.getMessage());
                status = 1;
            }
        }
        System.exit(status);
    }

}

File: lab_6/A3/ice_client.py
import Ice
import sys

slice_dir = Ice.getSliceDir()
if not slice_dir:
    print(sys.argv[0] + ': Slice directory not found.')
    sys.exit(1)

Ice.loadSlice("'-I" + slice_dir + "' slice/office.ice")


import Office


class ClientI(Office.Client):
    def registrationCallback(self, response, current=None):
        if response.success:
            print("SUCCESS: Received - Registration: ", response.carRegistration)
        else:
            print("FAILURE: Registration not completed")

    def instructorCallback(self, response, current=None):
        if response.success:
            print("SUCCESS: Received - Instructor: id: ", response.id)
        else:
            print("FAILURE: Instructor not completed")

    def licenseCallback(self, response, current=None):
        if response.success:
            print("SUCCESS: Received - License: id: ", response.id, "issuer: ", response.issuer)
        else:
            print("FAILURE: License not completed")


with Ice.initialize(sys.argv) as communicator:
    clerk = Office.ClerkPrx.checkedCast(communicator.stringToProxy('Clerk/Anetka : tcp -h 127.0.0.2 -p 10000 -z : udp -h 127.0.0.2 -p 10000 -z'))
    if not clerk:
        print("invalid proxy")
        sys.exit(1)

    #
    # Create an object adapter with no name and no endpoints for receiving callbacks
    # over bidirectional connections.
    #
    adapter = communicator.createObjectAdapter("")

    #
    # Register the callback receiver servant with the object adapter and activate
    # the adapter.
    #
    client_proxy = Office.ClientPrx.uncheckedCast(adapter.addWithUUID(ClientI()))
    adapter.activate()

    #
    # Associate the object adapter with the bidirectional connection.
    #
    clerk.ice_getConnection().setAdapter(adapter)

    #
    # Provide the proxy of the callback receiver object to the server and wait for
    # shutdown.
    #
    data = Office.PersonData("Arkadiusz", "Cwikla", "505abc")
    clerk.sayHello(data, client_proxy)

    while (True):
        text = input("==>")
        if text == "reg":
            time = clerk.register(data, Office.RegistrationRequest("Opel", 2003,
                                  Office.engineType.GAS))
            print("Time to wait: ", time)
        elif text == "ren":
            categories = [Office.licenseCategory.A, Office.licenseCategory.B]
            time = clerk.renewLicense(data, Office.RenewLicenseRequest(True, 5, categories))
            print("Time to wait: ", time)
        elif text == "ren-bad":
            categories = [Office.licenseCategory.A, Office.licenseCategory.B]
            time = clerk.renewLicense(data, Office.RenewLicenseRequest(False, 5, categories))
            print("Time to wait: ", time)
        elif text == "ins":
            time = clerk.instructorLicense(data, Office.InstructorLicenseRequest(True, True, True))
            print("Time to wait: ", time)
        elif text == "ins-bad":
            time = clerk.instructorLicense(data, Office.InstructorLicenseRequest(True, False, True))
            print("Time to wait: ", time)
        elif text == "x":
            break

    communicator.waitForShutdown()
